proposals/ref-readonly-parameters.md

```diff
@@ -0,0 +1,104 @@
+# `ref readonly` parameters
+
+## Summary
+[summary]: #summary
+
+Allow parameter declaration-site modifier `ref readonly` and change call-site rules as follows:
+
+| Call site annotation | `ref` parameter | `ref readonly` parameter | `in` parameter |
+|----------------------|-----------------|--------------------------|----------------|
+| `ref`                | Allowed         | Allowed                  | **Warning**    |
+| `in`                 | Error           | Allowed                  | Allowed        |
+| No annotation        | Error           | Warning                  | Allowed        |
+
+- `ref` and `ref readonly` parameters require lvalues at the call site.
+- `in` allows lvalues or rvalues at the callsite.
+
+Note that there is one change (in bold) to the existing rules which are currently as follows:
+
+| Call site annotation | `ref` parameter | `in` parameter |
+|----------------------|-----------------|----------------|
+| `ref`                | Allowed         | **Error**      |
+| `in`                 | Error           | Allowed        |
+| No annotation        | Error           | Allowed        |
```

Comment:
Need to cover how `ref readony` impacts implements, override and inherits scenarios. Essentially is the following legal or not?

```csharp
class C1
{
    protected virtual void M(in int i) { }
}

class C2 : C1
{
    protected override void M(ref readonly int i) { }
}
```

I do not believe we covered this in LDM. My recommendation is that the it be legal for `in` and `ref readonly` to be interchangable in these scenarios. Otherwise it cuts against the motivation of the proposal here.

Reply:
Here's my 2 cents: `in` allows both lvalues and rvalues, whereas `ref readonly` only allows lvalues (without warning). I would think that if you override a function which has an `in` parameter, with one that has a `ref readonly` parameter, you should get a warning, since you're saying "I don't want this kind of input", in a way that is invisible to anyone using the base class. I do think either should be overridable in either way though.

Effectively what I'm suggesting:
```csharp
class C1
{
    protected virtual void M(in int i) { }
}

class C2 : C1
{
    //warning (at definition of function) as some inputs (rvalues) are not intended for this function, make sure that the person defining this function knows that people can still give rvalues anyway (without a warning) through the base class.
    protected override void M(ref readonly int i) { }
}

class C3 : C2
{
    //no warning since this expands the possible options for input from just lvalues, to lvalues or rvalues.
    protected override void M(in int i) { }
}
```